src/start_server.ts

```diff
@@ -44,33 +53,59 @@ export interface ServerOptions {
 
   /** The package name to use for the root directory **/
   packageName?: string;
+
+  /** The HTTP protocol to use */
+  protocol?: string;
+
+  /** Path to TLS service key for HTTPS */
+  keyPath?: string;
+
+  /** Path to TLS certificate for HTTPS */
+  certPath?: string;
+
+  /** Path to H2 push-manifest file */
+  pushManifestPath?: string;
 }
 
-function applyDefaultOptions(options: ServerOptions): ServerOptions {
+async function applyDefaultOptions(options: ServerOptions): Promise<ServerOptions> {
   const withDefaults = Object.assign({}, options);
   Object.assign(withDefaults, {
-    port: options.port || 8080,
-    hostname: options.hostname || "localhost",
+    port: await nextOpenPort(options.port),
+    hostname: options.hostname || 'localhost',
     root: path.resolve(options.root || '.'),
+    certPath: options.certPath || 'cert.pem',
+    keyPath: options.keyPath || 'key.pem',
   });
   return withDefaults;
 }
 
 /**
- * @return {Promise} A Promise that completes when the server has started.
+ * If port unspecified/negative, finds an open port on localhost
+ * @param {number} port
+ * @returns {Promise<number>} Promise of open port
  */
-export function startServer(options: ServerOptions): Promise<http.Server> {
-  return new Promise((resolve, reject) => {
-    options = options || {};
-    if (options.port) {
-      resolve(options);
-    } else {
-      findPort(8080, 8180, (ports) => {
-        options.port = ports[0];
-        resolve(options);
+async function nextOpenPort(port: number): Promise<number> {
+  if (!port || port < 0) {
+    port = await new Promise<number>(resolve => {
+      findPort(8080, 8180, (ports: number[]) => {
+        resolve(ports[0]);
```

Comment:
It'd be nice if this function threw if no ports were unavailable.

Something like:

```typescript
let ports = [8080, 8180];
if (!port || port < 0) {
  ports = [port];
}
return new Promise<number>((resolve, reject) => {
  findPort(ports, (availablePorts: number[]) => {
    if (availablePorts.length > 0) {
      resolve(ports[0]);
      return;
    }
    reject(new Error(`No available ports for http server to bind to. Tried: ${ports.join(', ')}`));
  });
});

```